### ISSUE TITLE
论文复现：nn.Embedding无法前向对齐

### ISSUE THREAD

**Issue by akari0216** [PaddlePaddle/Paddle#38189]
paddlepaddle版本：2.2.1
pytorch版本：1.10.0+cu113

torch.nn.Embedding和paddle.nn.Embedding在前向过程中，有一定概率出现diff
代码如下：
---------------
import torch
import paddle
import numpy as np

#构造embedding
torch_enc = torch.nn.Embedding(10,7,padding_idx=1)
paddle_enc = paddle.nn.Embedding(10,7,padding_idx=1)

#构造权重并载入
emb_data = np.random.randn(10,7).astype(np.float32)
torch_emb = torch.tensor(emb_data)
paddle_emb = paddle.to_tensor(emb_data)

torch_enc.load_state_dict({"weight":torch_emb})
paddle_enc.load_dict({"weight":paddle_emb})

torch_enc.eval()
paddle_enc.eval()

#构造fake data
fake_data = np.random.randint(0,10,size=(3,2))
torch_x = torch.tensor(fake_data,dtype=torch.int64)
paddle_x = paddle.to_tensor(fake_data,dtype=paddle.int64)

torch_y = torch_enc(torch_x)
paddle_y = paddle_enc(paddle_x)

#打印diff
diff = np.allclose(torch_y.cpu().detach().numpy(),paddle_y.detach().numpy())
diff2 = np.mean(np.abs(torch_y.cpu().detach().numpy() - paddle_y.detach().numpy()))
print(diff)
print(diff2)

---------------
反复运行以上部分代码，会有相当概率打印出diff对不齐的情况，且diff2的值偏大

**Comment by joey12300** [comment]
@akari0216  试试在cpu也跑跑看？

**Comment by akari0216** [comment]
> @akari0216 试试在cpu也跑跑看？

在cpu环境下同样也出现了一定概率diff对不齐的情况
paddlepaddle版本：2.1.3
pytorch版本：1.10.0+cpu

**Comment by joey12300** [comment]
@akari0216 嗯嗯我稍晚点也试试看。你可以先试试paddle 2.2.1有没有相同问题

**Comment by akari0216** [comment]
> @akari0216 嗯嗯我稍晚点也试试看。你可以先试试paddle 2.2.1有没有相同问题

已在cpu环境下paddle 2.2.1 和pytorch 1.10.0+cpu进行测试，仍然出现该问题

**Comment by joey12300** [comment]
![image](https://user-images.githubusercontent.com/10826371/146507922-3f1bbba7-eceb-4916-9ab9-b4d8eb6f00db.png)

@akari0216 你好。这里主要是Padding的原因。torch在构造embedding的时候指定padding_idx,对应的参数会设置为全0；一旦给embedding赋值，就会覆盖掉padding_idx对应的0参数；而Paddle一旦指定了padding_idx，无论后续是否赋值，padding对应参数均为0.

**Comment by akari0216** [comment]
> ![image](https://user-images.githubusercontent.com/10826371/146507922-3f1bbba7-eceb-4916-9ab9-b4d8eb6f00db.png)
> 
> @akari0216 你好。这里主要是Padding的原因。torch在构造embedding的时候指定padding_idx,对应的参数会设置为全0；一旦给embedding赋值，就会覆盖掉padding_idx对应的0参数；而Paddle一旦指定了padding_idx，无论后续是否赋值，padding对应参数均为0.

感谢解答~那请问如果需要设定padding_idx参数的时候，有什么方案可以使得两者不会有概率出现diff的问题呢？

**Comment by joey12300** [comment]
@akari0216 要是你在构造参数时，将padding_idx对应的embedding设置为全0，就不会出现diff了。

`emb_data[1,:] = 0.0`

**Comment by akari0216** [comment]
> @akari0216 要是你在构造参数时，将padding_idx对应的embedding设置为全0，就不会出现diff了。
> 
> `emb_data[1,:] = 0.0`

感谢，按此方法已经对齐
